docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,108 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to be able to always identify the tenant correctly, queries using multiple tenants should inject a tenant label named `__tenant__` and its value containing the tenant ID into the results. A potentially existing label with the same name should be stored in a label with the prefix `original_`.
```

Comment:
> queries using multiple tenants should inject a tenant label named `__tenant__`

In the blocks storage we already have this:

```
TenantIDExternalLabel = "__org_id__"
```

It's an unfortunate label name we picked before deciding to standardise on "tenant ID", but I would try to aim for consistency. We should be able to move the blocks storage to the new label name, but I would suggest to use `__tenant_id__` instead of `__tenant__`. The naming `__tenant_id__` is more flexible because leaves the `__tenant` prefix open for future additional internal labels (eg. `__tenant_<something else>_`).

Reply:
Good point and :crossed_fingers: we get it all consistent one day

Reply:
Once you introduce `__tenant_id__` I will take care of migrating to it in the blocks storage too.